internal/controller/preservation/helper.go

```diff
@@ -0,0 +1,305 @@
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package preservation
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"time"
+
+	"go.uber.org/zap"
+
+	"golang.org/x/exp/maps"
+
+	corev1 "k8s.io/api/core/v1"
+	k8serrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/errors"
+
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// generateName returns a backup name by appending the reinstall generation to the base name.
+func generateName(name, generation string) string {
+	return fmt.Sprintf("%s-%s", name, generation)
+}
+
+// addPreservedLabel adds a timestamp backup label to the given labels map to indicate preservation.
+func addPreservedLabel(labels map[string]string) map[string]string {
+	if labels == nil {
+		labels = make(map[string]string)
+	}
+	labels[preservedLabel] = metav1.NewTime(time.Now()).String()
+	return labels
+}
+
+// removePreservedLabel removes the backup preservation label from the provided labels map.
+// If the labels map is nil, it returns the map unchanged.
+func removePreservedLabel(labels map[string]string) map[string]string {
+	if labels == nil {
+		return labels
+	}
+	delete(labels, preservedLabel)
+	return labels
+}
+
+// backupConfigMap backs up a Kubernetes ConfigMap by preserving its metadata and data.
+// The backup ConfigMap includes a preservation label and is created with the specified name.
+func backupConfigMap(
+	ctx context.Context,
+	c client.Client,
+	namespace, name, preservedName string,
+) error {
+	// Fetch the original ConfigMap.
+	originalConfigMap := &corev1.ConfigMap{}
+	if err := c.Get(ctx, types.NamespacedName{Namespace: namespace, Name: name}, originalConfigMap); err != nil {
+		return fmt.Errorf("failed to retrieve ConfigMap %s/%s for preservation: %w", namespace, name, err)
+	}
+
+	// Marshal the ConfigMap data into JSON.
+	encodedData, err := json.Marshal(originalConfigMap.Data)
+	if err != nil {
+		return fmt.Errorf("failed to marshal ConfigMap.Data for preservation: %w", err)
+	}
+
+	// Create the backup ConfigMap with the specified preserved name.
+	// - The ObjectMeta includes the original annotations and labels, along with an additional label indicating
+	//   preservation.
+	// - The Data field uses the original resource name as the key and stores the marshaled data as the value.
+	//   This approach ensures that the original ConfigMap is not modified or deleted and can be restored later if
```

Comment:
Why not put the entire configmap here?

This help ensure the data isn't edited, but why store the labels and annotations in the backup configmap labels and annotations?